diff --git a/drivers/misc/1 b/drivers/misc/1
new file mode 100644
index 000000000000..25da7b7b742d
--- /dev/null
+++ b/drivers/misc/1
@@ -0,0 +1,438 @@
+#include <kunit/test.h>
+#include <linux/slab.h>
+#include <linux/gfp.h>
+#include <linux/highmem.h>
+#include <linux/kthread.h>
+#include <linux/semaphore.h>
+#include <linux/spinlock_types.h>
+#include <linux/mutex.h>
+#include <linux/delay.h>
+#include "example.h"
+
+	/* Define the test cases. */
+static void misc_example_add_test_basic(struct kunit *test)
+{
+	KUNIT_EXPECT_EQ(test, 1, misc_example_add(1, 0));
+	KUNIT_EXPECT_EQ(test, 2, misc_example_add(1, 1));
+	KUNIT_EXPECT_EQ(test, 0, misc_example_add(-1, 1));
+	KUNIT_EXPECT_EQ(test, INT_MAX, misc_example_add(0, INT_MAX));
+	KUNIT_EXPECT_EQ(test, -1, misc_example_add(INT_MAX, INT_MIN));
+}
+
+static void test_print_char_size(struct kunit *test)
+{
+	KUNIT_EXPECT_EQ(test, 9, print_char_size());
+}
+
+static void test_print_char(struct kunit *test)
+{
+	char *s = print_char();
+	KUNIT_EXPECT_EQ(test, 't', *s);
+}
+
+
+static void test_print_char_input(struct kunit *test)
+{
+	char *s = print_char_input("testing");
+	KUNIT_EXPECT_EQ(test, 't', *s);
+}
+
+
+static void test_compare_char_each(struct kunit *test)
+{
+	char * s1 = "testing\n";
+	char * s2 = print_char();
+	KUNIT_EXPECT_EQ(test, 0, char_arr_cmp(s1,s2));
+}
+
+static void test_compare_char_each_input(struct kunit *test)
+{
+	char * s1 = "testing\n";
+	char * s2 = print_char_input("testing\n");
+	KUNIT_EXPECT_EQ(test, 0, char_arr_cmp(s1,s2));
+}
+
+
+static void misc_example_test_failure(struct kunit *test)
+{
+	KUNIT_FAIL(test, "This test never passes.");
+}
+
+static void example_test_allocation(struct kunit *test)
+{
+                char *buffer = kunit_kzalloc(test, 16, GFP_KERNEL);
+                /* Ensure allocation succeeded. */
+                KUNIT_ASSERT_NOT_ERR_OR_NULL(test, buffer);
+
+                KUNIT_ASSERT_STREQ(test, buffer, "");
+}
+
+static void list_del_init_test(struct kunit *test)
+{
+	struct list_head a, b;
+	LIST_HEAD(list);
+	
+	list_add_tail(&a, &list);
+	list_add_tail(&b, &list);
+
+        //before: [list] ->a ->b
+	list_del_init(&a);
+ 	
+	//after: [list] ->b, a initialized
+	KUNIT_EXPECT_EQ(test, list.next, &b);
+	KUNIT_EXPECT_EQ(test, b.prev, &list);
+	KUNIT_EXPECT_TRUE(test, list_empty_careful(&a));
+}
+
+static void test_strcmp(struct kunit *test)
+{
+	char * s1 = "testing\n";
+	char * s2 = print_char();
+	KUNIT_EXPECT_EQ(test, 0, strcmp(s1,s2));
+}
+
+static void test_compare_kalloc(struct kunit *test)
+{
+	void* k1 = kmalloc(16, GFP_KERNEL);
+	void* k2 = kmalloc(16, GFP_KERNEL);
+	KUNIT_EXPECT_EQ(test, 0, compare_kalloc(k1,k2));
+}
+
+
+static void test_compare_pages(struct kunit *test)
+{
+
+	void* p1 = alloc_page(GFP_KERNEL);
+	void* p2 = alloc_page(GFP_KERNEL);
+	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
+}
+
+static void test_compare_pages_in_order(struct kunit *test)
+{
+	struct page *p1 = alloc_pages(GFP_KERNEL,1);
+	struct page *p2 = alloc_page(GFP_KERNEL);
+	struct list_head *lh = &p2->lru;
+	struct list_head *nh = lh->next;
+	//struct list_head *lh = p2->lru.next;
+
+	p2 = container_of(lh, struct page, lru);
+	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
+}
+
+
+static void test_compare_pages_in_order2(struct kunit *test)
+{
+	struct page *pages = alloc_pages(GFP_KERNEL,1);
+	struct page *p1;
+	p1 = kmap(pages);
+	struct page *p2;
+	struct list_head *lh;
+	lh = &pages->lru;
+	//struct list_head *nh = lh->next;
+	//struct list_head *nh = lh->next;
+	//struct list_head *lh = p2->lru.next;
+
+	//p2 = container_of(lh->next, struct page, lru);
+	p2 = p1 + PAGE_SIZE;
+	//printk(KERN_INFO "list: %p, list->next: %p",lh, lh->next);
+	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
+}
+
+
+static void test_compare_pages_in_order3(struct kunit *test)
+{
+	struct page *pages = alloc_pages(GFP_KERNEL,1);
+	struct page *p1;
+	p1 = kmap_to_page(&pages[0]);
+	struct page *p2;
+	p2 = kmap_to_page(&pages[1]);
+	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
+}
+
+static void test_compare_pages_in_order4(struct kunit *test)
+{
+	struct page *pages = alloc_pages(GFP_KERNEL,1);
+	struct page *p1;
+	p1 = kmap_to_page(&pages[0]);
+	struct page *px;
+	px = kmap_to_page(&pages[1]);
+	struct list_head *lh = &p1->lru;
+	struct list_head *nh = lh->next;
+	//struct list_head *lh = p2->lru.next;
+
+	struct page *p2;
+	p2 = container_of(nh, struct page, lru);
+	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
+}
+
+static void test_compare_pages_in_order5(struct kunit *test)
+{
+	struct page *pages = alloc_pages(GFP_KERNEL,1);
+	struct page *p1 = kmap_to_page(&pages[0]);// without this line there will be a mm error!
+	//struct page *px;
+	//px = kmap_to_page(&pages[1]);
+	struct list_head *lh = &p1->lru;
+	struct list_head *nh = lh->next;
+	//struct list_head *lh = p2->lru.next;
+
+	struct page *p2;
+	p2 = container_of(nh, struct page, lru);
+	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
+}
+
+
+
+static void test_mainthread(struct kunit *test){
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //task1 = kthread_run(&thread_func, k1, "thread_func_1");
+    //kthread_stop(task1);
+    thread_func(k1);
+    int *rst;
+    rst = (int *)k1;
+    KUNIT_EXPECT_EQ(test, 20, *rst);
+}
+static struct semaphore sem;
+int thread_atomic(void *data){
+    int *d;
+    d = (int *)data;
+    atomic_t n = ATOMIC_INIT(*d);
+    int i = 0;
+    while(i < 10){
+	atomic_inc(&n);
+        i++;
+    }
+    *d = atomic_read(&n);
+    printk("inside atomic : %d\n", *d);
+    //*n = 20;
+    up(&sem);
+    return 0;
+    //do_exit();
+}
+
+int thread_funct(void *data){
+    int *n;
+    n = (int *)data;
+    int i = 0;
+    while(i < 10){
+	*n = *n + 1;
+        printk("%d\n", *n);
+        i++;
+    }
+    //*n = 20;
+    up(&sem);
+    return 0;
+    //do_exit();
+}
+
+static void test_kthread(struct kunit *test){
+    printk("Module starting ... ... ..\n");
+    printk("Module starting kthread test... ... ..\n");
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    sema_init(&sem,0);  
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //DEFINE_SEMAPHORE(&sem);
+    task1 = kthread_run(&thread_funct, k1, "thread_func_1");
+    down(&sem);// wait until 'up(&sem) in thread_atomic' triggered
+    	       // otherwise, mainthread could finished earlier then kthread
+    //thread_func(k1);
+    int *rst;
+    rst = (int *)k1;
+    kthread_stop(task1);
+    KUNIT_EXPECT_EQ(test, 20, *rst);
+}
+
+static void test_kthread2(struct kunit *test){
+    printk("Module starting ... ... ..\n");
+    printk("Module starting kthread test... ... ..\n");
+    struct task_struct *task1;
+    struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    void* k2 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    int var2 = 10;
+    k1 = &var;
+    k2 = &var2;
+    sema_init(&sem,0);  
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //DEFINE_SEMAPHORE(&sem);
+    task1 = kthread_run(&thread_funct, k1, "thread_func_1");
+    task2 = kthread_run(&thread_funct, k2, "thread_func_2");
+    down(&sem);// can only down(acquire) when other thread up(release)ed
+    down(&sem);// wait until 'up(&sem) in thread_atomic' triggered
+    	       // otherwise, mainthread could finished earlier then kthread
+    //thread_func(k1);
+    int *rst1;
+    rst1 = (int *)k1;
+    int *rst2;
+    rst2 = (int *)k2;
+    //kthread_stop(task1);
+    //kthread_stop(task2);
+    KUNIT_EXPECT_EQ(test, 20, *rst1);
+    KUNIT_EXPECT_EQ(test, 20, *rst2);
+}
+
+static void test_kthread3(struct kunit *test){
+    printk("Module starting ... ... ..\n");
+    printk("Module starting kthread test... ... ..\n");
+    struct task_struct *task1;
+    struct task_struct *task2;
+    volatile void* k1 = kmalloc(4, GFP_KERNEL);
+    volatile int var = 10;
+    k1 = &var;
+    sema_init(&sem,0);  
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //DEFINE_SEMAPHORE(&sem);
+    task1 = kthread_run(&thread_atomic, k1, "thread_func_1");
+    task2 = kthread_run(&thread_atomic, k1, "thread_func_2");
+    down(&sem);
+    down(&sem);// wait until 'up(&sem) in thread_atomic' triggered
+    	       // otherwise, mainthread could finished earlier then kthread
+    //thread_func(k1);
+    volatile int *rst;
+    rst = (int *)k1;
+    //kthread_stop(task1);
+    //kthread_stop(task2);
+    KUNIT_EXPECT_EQ(test, 30, *rst);
+}
+
+static void test_racethread(struct kunit *test){
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    task1 = kthread_run(&thread_func, k1, "thread_func_1");
+    //kthread_stop(task1);
+    //thread_func(k1);
+    int *rst;
+    rst = (int *)k1;
+    KUNIT_EXPECT_GT(test, 20, *rst);
+}
+
+
+spinlock_t my_lock;
+DEFINE_SPINLOCK(my_lock);
+
+int thread_func_spl(void *data){
+    spin_lock(&my_lock);
+    int *n;
+    n = (int *)data;
+    int i = 0;
+    while(i < 10){
+	*n = *n + 1;
+        printk("%d\n", *n);
+        i++;
+    }
+    //*n = 20;
+    spin_unlock(&my_lock);
+    return 0;
+    //do_exit();
+}
+
+static void test_splthread(struct kunit *test){
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    task1 = kthread_run(&thread_func_spl, k1, "thread_func_1");
+    //kthread_stop(task1);
+    //thread_func(k1);
+
+    msleep(1);//explicitly wait fot thread to take lock
+    spin_trylock(&my_lock);
+    int *rst;
+    rst = (int *)k1;
+    spin_unlock(&my_lock);
+    KUNIT_EXPECT_EQ(test, 20, *rst);
+}
+
+
+
+static void test_splthread2(struct kunit *test){
+    struct task_struct *task1;
+    struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    task1 = kthread_run(&thread_func_spl, k1, "thread_func_1");
+    task2 = kthread_run(&thread_func_spl, k1, "thread_func_2");
+    //kthread_stop(task1);
+    //thread_func(k1);
+
+    msleep(1);//explicitly wait fot thread to take lock
+    spin_trylock(&my_lock);
+    int *rst;
+    rst = (int *)k1;
+    spin_unlock(&my_lock);
+    KUNIT_EXPECT_EQ(test, 30, *rst);
+}
+
+struct mutex my_mutex;
+DEFINE_MUTEX(my_mutex);
+
+static void test_mutexhread(struct kunit *test){
+    struct task_struct *task1;
+    struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    mutex_init(&my_mutex);
+    task1 = kthread_run(&thread_func_spl, k1, "thread_func_1");
+    task2 = kthread_run(&thread_func_spl, k1, "thread_func_2");
+    //kthread_stop(task1);
+    //thread_func(k1);
+
+    //msleep(1);//explicitly wait fot thread to take lock
+    spin_trylock(&my_lock);
+    int *rst;
+    rst = (int *)k1;
+    spin_unlock(&my_lock);
+    KUNIT_EXPECT_EQ(test, 30, *rst);
+}
+
+static struct kunit_case misc_example_test_cases[] = {
+	KUNIT_CASE(misc_example_add_test_basic),
+	//KUNIT_CASE(misc_example_test_failure),
+	KUNIT_CASE(example_test_allocation),
+	KUNIT_CASE(test_print_char_size),
+	KUNIT_CASE(test_print_char_input),
+	KUNIT_CASE(test_print_char),
+	KUNIT_CASE(test_compare_char_each),
+	KUNIT_CASE(test_compare_char_each_input),
+	KUNIT_CASE(list_del_init_test),
+	KUNIT_CASE(test_strcmp),
+	KUNIT_CASE(test_compare_kalloc),
+	KUNIT_CASE(test_compare_pages),
+	KUNIT_CASE(test_compare_pages_in_order),
+	KUNIT_CASE(test_compare_pages_in_order2),
+	KUNIT_CASE(test_compare_pages_in_order3),
+	KUNIT_CASE(test_compare_pages_in_order4),
+	KUNIT_CASE(test_compare_pages_in_order5),
+	KUNIT_CASE(test_mainthread),
+	KUNIT_CASE(test_kthread),
+	KUNIT_CASE(test_kthread2),
+	KUNIT_CASE(test_kthread3),
+	KUNIT_CASE(test_racethread),
+	KUNIT_CASE(test_splthread),
+	KUNIT_CASE(test_splthread2),
+	{}
+};
+
+
+
+static struct kunit_suite misc_example_test_suite = {
+	.name = "misc-example",
+	.test_cases = misc_example_test_cases,
+};
+kunit_test_suite(misc_example_test_suite);
+
diff --git a/drivers/misc/example.c b/drivers/misc/example.c
index cb7be87cc70f..7ed595341509 100644
--- a/drivers/misc/example.c
+++ b/drivers/misc/example.c
@@ -1,6 +1,7 @@
 #include <linux/errno.h>
 #include <linux/slab.h>
 #include <linux/gfp.h>
+#include <linux/types.h>
 #include "example.h"
 
 int misc_example_add(int left, int right)
@@ -48,3 +49,20 @@ int char_arr_cmp(const char* s1,const char* s2)
     return *(const unsigned char*)s1-*(const unsigned char*)s2;
 }
 
+
+int thread_func(void *data){
+    int *n;
+    n = (int *)data;
+    int i = 0;
+    while(i < 10){
+	*n = *n + 1;
+        printk("%d\n", *n);
+        i++;
+    }
+    //*n = 20;
+    return 0;
+    //do_exit();
+}
+
+
+
diff --git a/drivers/misc/example.h b/drivers/misc/example.h
index 7cbbc2b2c99a..08d19a4e5596 100644
--- a/drivers/misc/example.h
+++ b/drivers/misc/example.h
@@ -5,3 +5,5 @@ int char_arr_cmp(const char* s1, const char* s2);
 char* print_char_input(char* input);//char* is almost the same as char[]
 int compare_kalloc(void* km1, void* km2);
 int compare_pages(struct page* p1,struct page* p2);
+int thread_func(void *data);
+int thread_atomic(void *data);
diff --git a/drivers/misc/example_test.c b/drivers/misc/example_test.c
index d2a69183fad9..4f4a3eae24d0 100644
--- a/drivers/misc/example_test.c
+++ b/drivers/misc/example_test.c
@@ -2,6 +2,11 @@
 #include <linux/slab.h>
 #include <linux/gfp.h>
 #include <linux/highmem.h>
+#include <linux/kthread.h>
+#include <linux/semaphore.h>
+#include <linux/spinlock_types.h>
+#include <linux/mutex.h>
+#include <linux/delay.h>
 #include "example.h"
 
 	/* Define the test cases. */
@@ -160,6 +165,271 @@ static void test_compare_pages_in_order4(struct kunit *test)
 	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
 }
 
+static void test_compare_pages_in_order5(struct kunit *test)
+{
+	struct page *pages = alloc_pages(GFP_KERNEL,1);
+	struct page *p1 = kmap_to_page(&pages[0]);// without this line there will be a mm error!
+	//struct page *px;
+	//px = kmap_to_page(&pages[1]);
+	struct list_head *lh = &p1->lru;
+	struct list_head *nh = lh->next;
+	//struct list_head *lh = p2->lru.next;
+
+	struct page *p2;
+	p2 = container_of(nh, struct page, lru);
+	KUNIT_EXPECT_EQ(test, 0, compare_pages(p1,p2));
+}
+
+
+
+static void test_mainthread(struct kunit *test){
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //task1 = kthread_run(&thread_func, k1, "thread_func_1");
+    //kthread_stop(task1);
+    thread_func(k1);
+    int *rst;
+    rst = (int *)k1;
+    KUNIT_EXPECT_EQ(test, 20, *rst);
+}
+static struct semaphore sem;
+int thread_atomic(void *data){
+    int *d;
+    d = (int *)data;
+    atomic_t n = ATOMIC_INIT(*d);
+    int i = 0;
+    while(i < 10){
+	atomic_inc(&n);
+        i++;
+    }
+    *d = atomic_read(&n);
+    printk("inside atomic : %d\n", *d);
+    //*n = 20;
+    up(&sem);
+    return 0;
+    //do_exit();
+}
+
+int thread_funct(void *data){
+    int *n;
+    n = (int *)data;
+    int i = 0;
+    while(i < 10){
+	*n = *n + 1;
+        printk("%d\n", *n);
+        i++;
+    }
+    //*n = 20;
+    up(&sem);
+    return 0;
+    //do_exit();
+}
+
+static void test_kthread(struct kunit *test){
+    printk("Module starting ... ... ..\n");
+    printk("Module starting kthread test... ... ..\n");
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    sema_init(&sem,0);  
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //DEFINE_SEMAPHORE(&sem);
+    task1 = kthread_run(&thread_funct, k1, "thread_func_1");
+    down(&sem);// wait until 'up(&sem) in thread_atomic' triggered
+    	       // otherwise, mainthread could finished earlier then kthread
+    //thread_func(k1);
+    int *rst;
+    rst = (int *)k1;
+    kthread_stop(task1);
+    KUNIT_EXPECT_EQ(test, 20, *rst);
+}
+
+static void test_kthread2(struct kunit *test){
+    printk("Module starting ... ... ..\n");
+    printk("Module starting kthread test... ... ..\n");
+    struct task_struct *task1;
+    struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    void* k2 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    int var2 = 10;
+    k1 = &var;
+    k2 = &var2;
+    sema_init(&sem,0);  
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //DEFINE_SEMAPHORE(&sem);
+    task1 = kthread_run(&thread_funct, k1, "thread_func_1");
+    task2 = kthread_run(&thread_funct, k2, "thread_func_2");
+    down(&sem);// can only down(acquire) when other thread up(release)ed
+    down(&sem);// wait until 'up(&sem) in thread_atomic' triggered
+    	       // otherwise, mainthread could finished earlier then kthread
+    //thread_func(k1);
+    int *rst1;
+    rst1 = (int *)k1;
+    int *rst2;
+    rst2 = (int *)k2;
+    //kthread_stop(task1);
+    //kthread_stop(task2);
+    KUNIT_EXPECT_EQ(test, 20, *rst1);
+    KUNIT_EXPECT_EQ(test, 20, *rst2);
+}
+
+static void test_kthread3(struct kunit *test){
+    printk("Module starting ... ... ..\n");
+    printk("Module starting kthread test... ... ..\n");
+    struct task_struct *task1;
+    struct task_struct *task2;
+    volatile void* k1 = kmalloc(4, GFP_KERNEL);
+    volatile int var = 10;
+    k1 = &var;
+    sema_init(&sem,0);  
+    //https://stackoverflow.com/questions/22469233/global-variable-wont-change-in-other-thread-in-c
+    //DEFINE_SEMAPHORE(&sem);
+    task1 = kthread_run(&thread_atomic, k1, "thread_func_1");
+    task2 = kthread_run(&thread_atomic, k1, "thread_func_2");
+    down(&sem);
+    down(&sem);// wait until 'up(&sem) in thread_atomic' triggered
+    	       // otherwise, mainthread could finished earlier then kthread
+    //thread_func(k1);
+    volatile int *rst;
+    rst = (int *)k1;
+    //kthread_stop(task1);
+    //kthread_stop(task2);
+    KUNIT_EXPECT_EQ(test, 30, *rst);
+}
+
+static void test_racethread(struct kunit *test){
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    task1 = kthread_run(&thread_func, k1, "thread_func_1");
+    //kthread_stop(task1);
+    //thread_func(k1);
+    int *rst;
+    rst = (int *)k1;
+    KUNIT_EXPECT_GT(test, 20, *rst);
+}
+
+
+spinlock_t my_lock;
+DEFINE_SPINLOCK(my_lock);
+
+int thread_func_spl(void *data){
+    spin_lock(&my_lock);
+    int *n;
+    n = (int *)data;
+    int i = 0;
+    while(i < 10){
+	*n = *n + 1;
+        printk("%d\n", *n);
+        i++;
+    }
+    //*n = 20;
+    spin_unlock(&my_lock);
+    return 0;
+    //do_exit();
+}
+
+static void test_splthread(struct kunit *test){
+    struct task_struct *task1;
+    //struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    task1 = kthread_run(&thread_func_spl, k1, "thread_func_1");
+    //kthread_stop(task1);
+    //thread_func(k1);
+
+    msleep(1);//explicitly wait fot thread to take lock
+    spin_trylock(&my_lock);
+    int *rst;
+    rst = (int *)k1;
+    spin_unlock(&my_lock);
+    KUNIT_EXPECT_EQ(test, 20, *rst);
+}
+
+
+
+static void test_splthread2(struct kunit *test){
+    struct task_struct *task1;
+    struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    task1 = kthread_run(&thread_func_spl, k1, "thread_func_1");
+    task2 = kthread_run(&thread_func_spl, k1, "thread_func_2");
+    //kthread_stop(task1);
+    //thread_func(k1);
+    msleep(1);//explicitly wait fot thread to take lock
+    spin_trylock(&my_lock);
+    int *rst;
+    rst = (int *)k1;
+    spin_unlock(&my_lock);
+    KUNIT_EXPECT_EQ(test, 30, *rst);
+}
+
+static void test_splthread3(struct kunit *test){
+    struct task_struct *task1;
+    struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    task1 = kthread_run(&thread_func_spl, k1, "thread_func_1");
+    task2 = kthread_run(&thread_func_spl, k1, "thread_func_2");
+    //kthread_stop(task1);
+    //thread_func(k1);
+    int *rst;
+    while(*rst != 30){
+        rst = (int *)k1;
+        msleep(1);// core will switch to another thread
+    }
+    KUNIT_EXPECT_EQ(test, 30, *rst);
+}
+
+struct mutex my_mutex;
+DEFINE_MUTEX(my_mutex);
+
+int thread_func_mutex(void *data){
+    mutex_lock(&my_mutex);
+    int *n;
+    n = (int *)data;
+    int i = 0;
+    while(i < 10){
+	*n = *n + 1;
+        printk("%d\n", *n);
+        i++;
+    }
+    //*n = 20;
+    mutex_unlock(&my_mutex);
+    return 0;
+    //do_exit();
+}
+
+static void test_mutexthread(struct kunit *test){
+    struct task_struct *task1;
+    struct task_struct *task2;
+    void* k1 = kmalloc(4, GFP_KERNEL);
+    int var = 10;
+    k1 = &var;
+    mutex_init(&my_mutex);
+    task1 = kthread_run(&thread_func_mutex, k1, "thread_func_1");
+    task2 = kthread_run(&thread_func_mutex, k1, "thread_func_2");
+    int *rst;
+    while(*rst != 30){
+        rst = (int *)k1;
+        msleep(1);// core will switch to another thread
+    }
+    KUNIT_EXPECT_EQ(test, 30, *rst);
+}
+
 static struct kunit_case misc_example_test_cases[] = {
 	KUNIT_CASE(misc_example_add_test_basic),
 	//KUNIT_CASE(misc_example_test_failure),
@@ -177,9 +447,21 @@ static struct kunit_case misc_example_test_cases[] = {
 	KUNIT_CASE(test_compare_pages_in_order2),
 	KUNIT_CASE(test_compare_pages_in_order3),
 	KUNIT_CASE(test_compare_pages_in_order4),
+	KUNIT_CASE(test_compare_pages_in_order5),
+	KUNIT_CASE(test_mainthread),
+	KUNIT_CASE(test_kthread),
+	KUNIT_CASE(test_kthread2),
+	KUNIT_CASE(test_kthread3),
+	KUNIT_CASE(test_racethread),
+	KUNIT_CASE(test_splthread),
+	KUNIT_CASE(test_splthread2),
+	KUNIT_CASE(test_splthread3),
+	KUNIT_CASE(test_mutexthread),
 	{}
 };
 
+
+
 static struct kunit_suite misc_example_test_suite = {
 	.name = "misc-example",
 	.test_cases = misc_example_test_cases,
